src/audio/copier/copier.c

```diff
@@ -266,6 +266,26 @@ static int create_host(struct comp_dev *parent_dev, struct copier_data *cd,
 	return ret;
 }
 
+static enum sof_ipc_stream_direction
+	get_gateway_direction(enum ipc4_connector_node_id_type node_id_type)
+{
+	/* WARNING: simple "% 2" formula that was used before does not work for all
+	 * interfaces: at least it does not work for IPC gateway. But it may also
+	 * does not work for other not yet supported interfaces. And so additional
+	 * "case"s might be required here in future.
+	 */
+	switch (node_id_type) {
+	/* from DSP to host */
+	case ipc4_ipc_output_class:
+		return SOF_IPC_STREAM_CAPTURE;
+	/* from host to DSP */
+	case ipc4_ipc_input_class:
+		return SOF_IPC_STREAM_PLAYBACK;
+	default:
+		return node_id_type % 2;
```

Comment:
humm, is this really needed?  I don't get what additional directions we might see beyond capture and playback.

And if you return an enum it should be from the enum list, not based on a formula.

Reply:
> humm, is this really needed? I don't get what additional directions we might see beyond capture and playback.
> 
> And if you return an enum it should be from the enum list, not based on a formula.

The "% 2" was not my invention, it was used before in original code. However, it worked incorrectly for ipc4_ipc_output_class and ipc4_ipc_input_class. Hence the formula was moved to a separate function with switch added to handle ipc4_ipc_output_class and ipc4_ipc_input_class separately.

Of course, its better to do not use such kind of formulas with enums but have a big switch and return enum value. However, with such correct approach I was not sure what to do with ipc4_gpio_class, ipc4_hda_host_inout_class, etc. What direction should be returned for them? We do not (yet?) use those gateway types. So I decided leave the formula for rest of the gateways as it was working before and added a comment that if the formula did not work properly for  ipc4_ipc_output_class and ipc4_ipc_input_class, it might also not work properly for some other gateways that will be enabled in future.